File: services/server/src/safe_socket/safe_socket.py
import socket

# TODO: Complete with a short-read/short-write tolerant implementation


def recv_all(socket: socket.socket, size):
    buff = b""
    while len(buff) < size:
        received = socket.recv(size - len(buff))
        buff += received

    return buff


def send_all(socket: socket.socket, bytes):
    total_sent = 0
    while total_sent < len(bytes):
        n = socket.send(bytes[total_sent:])
        total_sent += n
    return total_sent

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import (
	"io"
)

func SendAll(socket io.Writer, bytes []byte) error {
	totalSent := 0

	for totalSent < len(bytes) {
		n, err := socket.Write(bytes[totalSent:])
		if err != nil {
			return err
		}
		totalSent += n
	}

	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	buff := make([]byte, size)
	totalRead := 0

	for totalRead < size {
		n, err := socket.Read(buff[totalRead:])
		if err != nil {
			return nil, err
		}
		totalRead += n
	}

	return buff[:totalRead], nil
}
